packages/synthetix-main/contracts/storage/Distribution.sol

```diff
@@ -48,81 +146,121 @@ library Distribution {
-     * * amount of collateral (ex. when the total collateral amount might change over time)
-     * note: if you do not expect the global value to be scaled/modified, you might choose to store it as the shares in another crowdin distribution instead
+     * @dev Updates an actor's value in the distribution to the specified amount.
+     *
+     * The incoming value is translated into a number of shares,
+     * which modify the total number of shares in the distribution, and thus
+     * the distribution's totalValue.
+     *
+     * Returns the amount by which the distribution's number of shares changed.
+     *
+     * Note: A distribution is intended to either use updateActorShares or updateActorValue, but not both
+     * during the distribution's lifetime.
      */
     function updateActorValue(
         Data storage dist,
         bytes32 actorId,
         int value
     ) internal returns (uint shares) {
         if (dist.valuePerShare == 0 && dist.totalShares != 0) {
-            revert InvalidParameters("valuePerShare", "shares still exist when no value per share remains");
+            revert ZeroValuePerShare();
         }
 
+        DistributionActor.Data storage actor = dist.actorInfo[actorId];
+
+        if (actor.lastValuePerShare != 0) {
+            revert InconsistentDistribution();
+        }
+
+        // Calculate the number of shares that the
+        // change in value produces.
+
+        // If the distribution is empty, set valuePerShare to 1,
+        // and the number of shares to the given value.
         if (dist.totalShares == 0) {
             dist.valuePerShare = 1e27;
-            shares = uint(value > 0 ? value : -value);
-        } else {
+            shares = uint(value > 0 ? value : -value); // Ensure value is positive
+        }
+        // If the distribution is not empty, the number of shares
+        // is determined by the valuePerShare.
+        else {
+            // Calculate number of shares this way instead of value / valuePerShare,
+            // in order to avoid precision loss.
             shares = uint((value * int128(dist.totalShares)) / totalValue(dist));
         }
 
-        DistributionActor.Data storage actor = dist.actorInfo[actorId];
+        actor.shares = uint128(shares);
 
-        dist.totalShares = uint128(dist.totalShares + shares - actor.shares);
+        uint deltaActorShares = shares - actor.shares;
+        dist.totalShares += uint128(deltaActorShares);
 
-        actor.shares = uint128(shares);
-        // lastValuePerShare remains unset because they contributed
+        // Note: No need to udpate actor.lastValuePerShare
+        // because they contributed value to the distribution.
     }
 
     /**
-     * get the number of shares recorded for an actor
+     * @dev Upadates an actor's lastValuePerShare to the distribution's current valuePerShare, and
+     * returns the total change in value for the actor.
```

Comment:
returns the change in value from the actor from the last update

Reply:
Updated